### ISSUE TITLE
gtcrn转ncnn过程中遇到错误

### ISSUE THREAD

**Issue by alittleboy-c** [wxqwinner/gtcrn-ncnn#3]
大佬，你好。我最近也在研究https://github.com/Xiaobin-Rong/gtcrn，他里面有一个流模式的python文件https://github.com/Xiaobin-Rong/gtcrn/blob/main/stream/gtcrn_stream.py，我想把它转到ncnn推理框架下，用的腾讯pnnx转换工具，发现其中的class DPGRNN(nn.Module)中的self.inter_rnn模块转换失败，并且在尝试了很多方法后，还是没能解决这个问题，想问问大佬是通过什么方式转成功ncnn推理框架的，能不能给一些转ncnn相关的建议，非常感谢！！！！

**Comment by wxqwinner** [comment]
ncnn不支持batch维，而DPGRNN中batch维参与运算且不能简化掉，pnnx转模型的时候会试图简化掉，所以要重写相应的模块，你可以看param里那部分，自己实现GRU就好了

**Comment by alittleboy-c** [comment]
感谢大佬的建议。我用自己的GRU模块替换了nn.GRU模块，具体实现如下，但是单独试了自己实现的GRU转ncnn，没能成功，大佬能不能给些意见，或者方便开源一下手写GRU的实现吗，谢谢啦！！！
from typing import Tuple

import torch
from torch import nn, Tensor


class CustomGRUCell(nn.Module):
    def __init__(self, input_size: int, hidden_size: int):
        super().__init__()
        self.input_size = input_size
        self.hidden_size = hidden_size

        self.rz_gate = nn.Linear(input_size + hidden_size, hidden_size * 2)
        self.in_gate = nn.Linear(input_size, hidden_size)
        self.hn_gate = nn.Linear(hidden_size, hidden_size)
        ...

    def forward(self, xt: Tensor, ht0: Tensor):
        """
        xt: (batch_size, 1, input_size)
        ht0: (batch_size, hidden_size)
        """

        ht0 = ht0[:, None]  # (batch_size, 1, hidden_size)

        inputs = torch.cat([xt, ht0], -1)

        rz_gate_out = torch.sigmoid(self.rz_gate(inputs))
        r_gate_out, z_gate_out = torch.split(rz_gate_out, self.hidden_size, -1)  # Unsupported split axis !

        in_out = self.in_gate(xt)
        hn_out = self.hn_gate(ht0)
        n_gate_out = torch.tanh(in_out + r_gate_out * hn_out)

        ht = (1 - z_gate_out) * n_gate_out + z_gate_out * ht0
        return ht

    def set_weights(self, weight_ih_l0: Tensor, weight_hh_l0: Tensor, bias_ih_l0: Tensor, bias_hh_l0: Tensor):
        slice_index = self.hidden_size * 2
        rz_weight = torch.cat((weight_ih_l0[:slice_index], weight_hh_l0[:slice_index]), -1)
        rz_bias = bias_ih_l0[:slice_index] + bias_hh_l0[:slice_index]

        self.rz_gate.weight.data.copy_(rz_weight)
        self.rz_gate.bias.data.copy_(rz_bias)

        self.in_gate.weight.data.copy_(weight_ih_l0[slice_index:])
        self.in_gate.bias.data.copy_(bias_ih_l0[slice_index:])

        self.hn_gate.weight.data.copy_(weight_hh_l0[slice_index:])
        self.hn_gate.bias.data.copy_(bias_hh_l0[slice_index:])
        ...


class CustomGRU(nn.Module):
    def __init__(self, input_size: int, hidden_size: int, num_layers: int = 1):
        super().__init__()
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.num_layers = num_layers

        self.gru = nn.ModuleList(
            [CustomGRUCell(input_size if i == 0 else hidden_size, hidden_size) for i in range(num_layers)]
        )
        ...

    def forward(self, x: Tensor, states: Tensor) -> Tuple[Tensor, Tensor]:
        """
        x: (batch_size, 1, input_size)
        states: (num_layers, batch_size, hidden_size)
        """

        batch_size = x.shape[0]
        state_list = torch.unbind(states, dim=0)  # (batch_size, hidden_size)
        gru_out, h_states_list = x, []
        for i in range(self.num_layers):
            gru_out = self.gru[i](gru_out, state_list[i])  # (batch_size, 1, hidden_size)
            h_states_list.append(gru_out.view(1, batch_size, self.hidden_size))
        h_states = torch.cat(h_states_list, dim=0)
        return gru_out, h_states

    def set_weights(self, state_dict: dict):
        # initialize all weights and biases to zero
        for value in self.state_dict().values():
            value.data.zero_()

        keys = list(state_dict)
        for i in range(self.num_layers):
            self.gru[i].set_weights(*map(state_dict.get, keys[4 * i: 4 * (i + 1)]))

**Comment by wxqwinner** [comment]
如果你只是要现成的模型，仓库里有了，显然你不是，那么自己实现会比较好，不难

**Comment by alittleboy-c** [comment]
是的，我想整个过程都走一遍，可以系统的学习一下

**Comment by maoyapeng** [comment]
> 如果你只是要现成的模型，仓库里有了，显然你不是，那么自己实现会比较好，不难

王哥你好，GTCRN中除了batch维为并不是1维的这处GRU需要修改之外，还有哪些需要修改或者容易出错值得注意的点？希望王哥能够指点一下，十分感谢

**Comment by wxqwinner** [comment]
# self.inter_rnn = GRNN(
        #     input_size=input_size, hidden_size=hidden_size, bidirectional=False
        # )
        self.inter_rnn = SimpleGRNN(input_size=input_size, hidden_size=hidden_size)
        # self.inter_fc = nn.Linear(hidden_size, hidden_size)
        self.inter_fc = SimpleLinear(hidden_size, hidden_size)
        # self.inter_ln = nn.LayerNorm(((width, hidden_size)), eps=1e-8)
        self.inter_ln = SimpleLayerNorm(((width, hidden_size)), eps=1e-8)

**Comment by cp917** [comment]
SimpleGRNN SimpleLinear SimpleLinear 如何实现

**Comment by cp917** [comment]
@alittleboy-c 问一下你的转换是怎么做的

**Comment by jiayukk** [comment]
训练时可以不变，转换时可以将GRU按公式拆成全连接和激活，然后继承训练好的GRU权重。

**Comment by panissamarieglys512-alt** [comment]
> [@alittleboy-c](https://github.com/alittleboy-c) 问一下你的转换是怎么做的

我还没转成功，发现太麻烦了。

**Comment by panissamarieglys512-alt** [comment]
> 训练时可以不变，转换时可以将GRU按公式拆成全连接和激活，然后继承训练好的GRU权重。

你转换成功了吗，转成ncnn推理框架后，效果和转换前保持一致吗

**Comment by jiayukk** [comment]
完全一致，速度也是一样
